internal/volumehelper/volume_policy_helper.go

```diff
@@ -0,0 +1,243 @@
+	// if it is not snapshot then skip the code path for snapshotting the PV/PVC
+	if groupResource == kuberesource.PersistentVolumeClaims {
+		pvc := new(corev1api.PersistentVolumeClaim)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), &pvc); err != nil {
+			return false, err
+		}
+
+		if v.VolumePolicy != nil {
+			// fetch the PV for PVC and get matching action for it
+			pv, err := v.GetPVForPVC(pvc)
+			if err != nil {
+				return false, err
+			}
+			action, err := v.VolumePolicy.GetMatchAction(pv)
+			if err != nil {
+				return false, err
+			}
+
+			// Also account for SnapshotVolumes flag on backup CR
+			if action != nil && action.Type == resourcepolicies.Snapshot && boolptr.IsSetToTrue(v.SnapshotVolumes) {
+				return true, nil
+			}
+			return false, nil
+		}
+	}
+
+	if groupResource == kuberesource.PersistentVolumes {
+		pv := new(corev1api.PersistentVolume)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), &pv); err != nil {
+			return false, err
+		}
+
+		if v.VolumePolicy != nil {
+			action, err := v.VolumePolicy.GetMatchAction(pv)
+			if err != nil {
+				return false, err
+			}
+
+			// Also account for SnapshotVolumes flag on backup CR
+			if action != nil && action.Type == resourcepolicies.Snapshot && boolptr.IsSetToTrue(v.SnapshotVolumes) {
+				return true, nil
+			}
+			return false, nil
+		}
+	}
+
+	// now if volumepolicy is not specified then just check for snapshotVolumes flag
+	if boolptr.IsSetToTrue(v.SnapshotVolumes) {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+func (v *VolumeHelperImpl) GetVolumesForFSBackup(pod *corev1api.Pod) ([]string, []string, error) {
+	// Check if there is a fs-backup/snapshot volume policy specified by the user, if yes then use the volume policy approach to
+	// get the list volumes for fs-backup else go via the legacy annotation based approach
+	var includedVolumes, optedOutVolumes, volsToProcessByLegacyApproach = make([]string, 0), make([]string, 0), make([]string, 0)
+	var err error
+
+	if v.VolumePolicy != nil {
+		// Get the list of volumes to back up using pod volume backup for the given pod matching fs-backup volume policy action
+		v.Logger.Infof("Volume Policy specified by the user, using volume policy approach to segregate pod volumes for fs-backup")
+		includedVolumes, optedOutVolumes, volsToProcessByLegacyApproach, err = v.GetVolumesMatchingFSBackupAction(pod, v.VolumePolicy)
+		if err != nil {
+			return includedVolumes, optedOutVolumes, err
+		}
+	}
+
+	// process legacy annotation based approach, this will be done for 2 scenarios:
+	// 1. volume policy not specified by the user
+	// 2. there are some volumes for which the volume policy approach did not get any supported matching actions
+	if v.VolumePolicy == nil || len(volsToProcessByLegacyApproach) > 0 {
```

Comment:
if v.VolumePolicy == nil then len(volsToProcessByLegacyApproach) will always be 0, so should we remove the condition `len(volsToProcessByLegacyApproach) > 0`?